### PR TITLE
WSS-659 SecurityContextToken validator fixing QName

### PR DESCRIPTION
**Pull request by CauchyPeano**
Hi,

I am creating PR for following issue 
https://issues.apache.org/jira/projects/WSS/issues/WSS-659?filter=allopenissues

I would also like to add tests, but haven't found existing ones. 

Best
Igor

### CODE REVIEW

File: ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public void handle(InputProcessorChain inputProcessorChain, final XMLSecurityPro
         }
 
         final QName elementName = new QName(securityContextTokenTypeJAXBElement.getName().getNamespaceURI(),
-                WSSConstants.TAG_WSC0502_IDENTIFIER.getLocalPart());
+                securityContextTokenTypeJAXBElement.getName().getLocalPart());
```
Comment (coheigea) [inline ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java]: This fix will break the next line, which gets the identifier. Really "elementName" here should be "identifierElementName". So maybe rename elementName here to identifierElementName and revert the change. Then add a new elementName for the SecurityContextToken itself.

Comment (CauchyPeano) [reply ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java]: Yes, I see the problem. I just can't yet understand signature of `org.apache.wss4j.stax.validate.SecurityContextTokenValidator#validate` - what does identifier parameter do?

Comment (CauchyPeano) [reply ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java]: And do I also need to support two versions of NS:
"http://schemas.xmlsoap.org/ws/2005/02/sc" and "http://docs.oasis-open.org/ws-sx/ws-secureconversation/200512" ?

Comment (coheigea) [reply ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java]: The identifier is just a way of referring to a particular SecurityContextToken. No, you don't have to support two versions - all you need to do is to make sure that the security context token element is passed through to wssSecurityProperties.getValidator.

**Comment by coheigea** [comment]
@CauchyPeano are you going to pursue this PR? Otherwise I will fix it.

**Comment by CauchyPeano** [comment]
Hey @coheigea , yes, I want to do it. Was just in holiday coma.

**Comment by CauchyPeano** [comment]
Hey @coheigea ,

I am trying to write test now, and have difficulty to understanding how to write Validator. Here is what I am trying to do:

```
            AbstractInboundSecurityToken abstractInboundSecurityToken = new AbstractInboundSecurityToken(null, IDGenerator.generateID(null), SecurityTokenConstants.KeyIdentifier_KeyValue, true) {
                @Override
                public SecurityTokenConstants.TokenType getTokenType() {
                    return WSSecurityTokenConstants.SECURITY_CONTEXT_TOKEN;
                }

                @Override
                protected Key getKey(String algorithmURI, XMLSecurityConstants.AlgorithmUsage algorithmUsage, String correlationID) throws XMLSecurityException {
                    return new SecretKeySpec("test".getBytes(), "sha-256"); //I guess this part is wrong
                }
            };
            SecurityContextTokenValidator validator = new SecurityContextTokenValidator() {
                @Override
                public InboundSecurityToken validate(AbstractSecurityContextTokenType securityContextTokenType, String identifier, TokenContext tokenContext) {
                    System.out.println("Test!!");
                    return abstractInboundSecurityToken;
                }
            };
            securityProperties.addValidator(WSSConstants.TAG_WSC0502_SCT, validator);
            securityProperties.addValidator(WSSConstants.TAG_WSC0512_SCT, validator);
```

In this example I am getting exception at this line: `org/apache/santuario/xmlsec/2.1.4/xmlsec-2.1.4-sources.jar!/org/apache/xml/security/stax/impl/processor/input/AbstractSignatureInputHandler.java:354`

What will be the correct way to configure it for tests?

**Comment by coheigea** [comment]
Why are you adding a AbstractInboundSecurityToken ? If you look at say the test testSCTDKTEncryptInbound in SecurityContextTokenTest.java all you need to do is add the Validator there.

**Comment by CauchyPeano** [comment]
Yes, but can I return `null` in `validate` result?

**Comment by coheigea** [comment]
The easiest way is just to subclass SecurityContextTokenValidatorImpl and call super.validate

**Comment by CauchyPeano** [comment]
Thanks, it's working this way. I've updated PR

File: ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java [hunk 2]
```diff
@@ -108,9 +108,22 @@ public String getId() {
...
 
         //fire a tokenSecurityEvent
+        SecurityContextTokenSecurityEvent securityEvent = createTokenSecurityEvent(securityContextTokenType, securityTokenProvider);
+        wsInboundSecurityContext.registerSecurityEvent(securityEvent);
+    }
+
+    private QName extractElementName(JAXBElement<?> jaxbElement) {
+        QName qName = jaxbElement.getName();
+        return new QName(qName.getNamespaceURI(), qName.getLocalPart());
+    }
```
Comment (coheigea) [inline ws-security-stax/src/main/java/org/apache/wss4j/stax/impl/processor/input/SecurityContextTokenInputHandler.java]: This method can be removed, and instead just do:

final QName elementName = securityContextTokenTypeJAXBElement.getName();

File: ws-security-stax/src/test/java/org/apache/wss4j/stax/test/SecurityContextTokenTest.java [hunk 3]
```diff
@@ -1114,4 +1122,66 @@ public void testSCTSign() throws Exception {
...
+            final boolean[] validatorCalled = {false};
+            SecurityContextTokenValidator validator = new SecurityContextTokenValidatorImpl() {
+                @Override
+                public InboundSecurityToken validate(AbstractSecurityContextTokenType securityContextTokenType, String identifier, TokenContext tokenContext) throws WSSecurityException {
+                    validatorCalled[0] = true;
+                    return super.validate(securityContextTokenType, identifier, tokenContext);
+                }
+            };
+            securityProperties.addValidator(WSSConstants.TAG_WSC0502_SCT, validator);
+            securityProperties.addValidator(WSSConstants.TAG_WSC0512_SCT, validator);
```
Comment (coheigea) [inline ws-security-stax/src/test/java/org/apache/wss4j/stax/test/SecurityContextTokenTest.java]: Replace these two lines with the following, to make sure we are only setting the validator for the correct version:

if (version == ConversationConstants.VERSION_05_02) {
                securityProperties.addValidator(WSSConstants.TAG_WSC0502_SCT, validator);
            } else {
                securityProperties.addValidator(WSSConstants.TAG_WSC0512_SCT, validator);
            }

Comment (CauchyPeano) [reply ws-security-stax/src/test/java/org/apache/wss4j/stax/test/SecurityContextTokenTest.java]: done

**Comment by CauchyPeano** [comment]
Thanks @coheigea ,
maybe there is some other jira ticket with which I can help?

**Comment by coheigea** [comment]
Any help would be welcome - https://issues.apache.org/jira/projects/WSS/issues/WSS-638?filter=allopenissues
